core/src/main/java/io/questdb/griffin/engine/groupby/SampleByFillValueRecordCursorFactory.java

```diff
@@ -154,43 +178,54 @@ public void close() {
-        // aggregation results and timestamp
-
-        int n = groupByFunctions.size();
-        final Record baseCursorRecord = baseCursor.getRecord();
-        while (baseCursor.hasNext()) {
-            MapKey key = map.withKey();
-            mapSink.copy(baseCursorRecord, key);
-            MapValue value = key.createValue();
-            if (value.isNew()) {
-                // timestamp is always stored in value field 0
-                value.putLong(0, Numbers.LONG_NaN);
-                // have functions reset their columns to "zero" state
-                // this would set values for when keys are not found right away
-                for (int i = 0; i < n; i++) {
-                    groupByFunctions.getQuick(i).setNull(value);
+        try {
+            map.clear();
+            long maxInMemoryRows = executionContext.getCairoSecurityContext().getMaxInMemoryRows();
+            if (maxInMemoryRows > baseCursor.size() || baseCursor.size() < 0) {
+                map.setMaxSize(maxInMemoryRows);
+
+                // This factory fills gaps in data. To do that we
+                // have to know all possible key values. Essentially, every time
+                // we sample we return same set of key values with different
+                // aggregation results and timestamp
+
+                int n = groupByFunctions.size();
+                final Record baseCursorRecord = baseCursor.getRecord();
+                while (baseCursor.hasNext()) {
+                    MapKey key = map.withKey();
+                    mapSink.copy(baseCursorRecord, key);
+                    MapValue value = key.createValue();
+                    if (value.isNew()) {
+                        // timestamp is always stored in value field 0
+                        value.putLong(0, Numbers.LONG_NaN);
+                        // have functions reset their columns to "zero" state
+                        // this would set values for when keys are not found right away
+                        for (int i = 0; i < n; i++) {
+                            groupByFunctions.getQuick(i).setNull(value);
+                        }
+                    }
+                }
+
+                // empty map? this means that base cursor was empty
+                if (map.size() == 0) {
+                    baseCursor.close();
+                    return EmptyTableRecordCursor.INSTANCE;
                 }
-            }
-        }
 
-        // empty map? this means that base cursor was empty
-        if (map.size() == 0) {
+                // because we pass base cursor twice we have to go back to top
+                // for the second run
+                baseCursor.toTop();
+                boolean next = baseCursor.hasNext();
+                // we know base cursor has value
+                assert next;
+                return initFunctionsAndCursor(executionContext, baseCursor);
+            } else {
+                throw LimitOverflowException.instance(maxInMemoryRows);
```

Comment:
this is purely syntax, (no extra code should be generated). But when there is `return` statement inside `if` statement, `else` branch is not necessary